### PR TITLE
[PROCS-3602] Prevent collecting excessive number of zombie processes

### PR DESCRIPTION
**Pull request by levin0710**
### What does this PR do?

Adds a config setting to toggle collection of zombie processes on/off. Uses logic to determine if zombie process should be collected. The default behavior for this configuration is going to be false. This means that the zombie process will continue to collected as long as the user does not specify that they want to skip the collection of zombie processes.

### Motivation

We’ve run into OOMs several times that have been traced back to excessive numbers of zombie processes. This new config has the goal of preventing this issue by giving users the option of skipping the collection of zombie processes. 
This is tracked in Jira by this ticket: https://datadoghq.atlassian.net/browse/PROCS-3602
### Additional Notes

You will notice that in this PR the toggle for collecting zombie processes is being used both in `pkg/process/procutil/process_linux.go` and `pkg/process/checks/process.go`.  In theory, catching the zombie processes in `process_linux.go` should be enough to stop the agent from OOMing. However, it does not hurt to keep the logic in `process.go` as some sort of fail safe. Should we keep this?

### Possible Drawbacks / Trade-offs

The testing process for this change could be improved. Implementing e2e tests for zombie processes is something that could be explored in future changes to ensure correctness and functionality.

### Describe how to test/QA your changes

```
# datadog.yaml
process_config:
    ignore_zombie_processes: true
```

Using the above config, run the following steps:

- Generate artificial zombie processes with the following command `(sleep 1 & exec /bin/sleep 10)`
- Run the Process check
- Verify that no processes with `"state" : 7` (zombie) exist in the check output

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/process/checks/process.go [hunk 1]
```diff
@@ -137,6 +141,8 @@ func (p *ProcessCheck) Init(syscfg *SysProbeConfig, info *HostInfo, oneShot bool
 
 	p.disallowList = initDisallowList(p.config)
 
+	p.zombiesDisallowed = initZombieDisallowed(p.config)
```
Comment (daniel-taf) [inline pkg/process/checks/process.go:144]: No need to use a function here, its only 1 line set this.

File: pkg/process/checks/process.go [hunk 2]
```diff
@@ -678,3 +689,8 @@ func isDisallowListed(cmdline []string, disallowList []*regexp.Regexp) bool {
 	}
 	return false
 }
+
+func initZombieDisallowed(config ddconfig.Reader) bool {
+	disallowed := config.GetBool(configDisallowZombies)
+	return disallowed
+}
```
Comment (wiyu) [inline pkg/process/checks/process.go:696]: [minor] This function doesn't really do anything besides make a call to `getBool`. You should just inline it.

File: pkg/process/checks/process.go [hunk 3]
```diff
@@ -592,6 +600,9 @@ func skipProcess(
 		// This means short-lived processes (<2s) will never be captured.
 		return true
 	}
+	if zombiesDisallowed && fp.Stats.Status == "Z" {
```
[the remark this answers is not in the document]
Comment (daniel-taf) [reply pkg/process/checks/process.go:603]: Out of curiosity, what determines whether its nil or not?

[the remark this answers is not in the document]
Comment (levin0710) [reply pkg/process/checks/process.go:603]: In the same function there is some similar "fp.<field>" calls. Should i check fp is not nil before all other function logic?

File: pkg/config/setup/process_test.go [hunk 4]
```diff
@@ -304,6 +308,12 @@ func TestEnvVarOverride(t *testing.T) {
 			value:    "datacat.com",
 			expected: "datacat.com",
 		},
+		{
+			key:      "process_config.ignore_zombie_processes.enabled",
+			env:      "DD_PROCESS_CONFIG_IGNORE_ZOMBIE_PROCESSES",
+			value:    "false",
+			expected: false,
```
Comment (wiyu) [inline pkg/config/setup/process_test.go:315]: [bug] The env var key you're using here doesn't match the one you've set. Also this test is to verify setting a non-default value works. In this case that would mean setting and expecting it to be `true`.

**Review by wiyu — COMMENTED** [review]
Looks good so far. Let's also add some tests which asserts zombie process collection is skipped.

File: pkg/process/checks/process.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wiyu) [reply pkg/process/checks/process.go:603]: So it does look like we always populate/create the stats field, but its not obvious without looking through the code. I think you can just add a check in this call and leave the others as is to keep the blast radius of the change small.

File: pkg/process/checks/process.go [hunk 5]
```diff
@@ -36,6 +36,7 @@ const (
 	configScrubArgs            = configPrefix + "scrub_args"
 	configStripProcArgs        = configPrefix + "strip_proc_arguments"
 	configDisallowList         = configPrefix + "blacklist_patterns"
+	configDisallowZombies      = configPrefix + "ignore_zombie_processes"
```
Comment (robertjli) [inline pkg/process/checks/process.go:39]: Let's be consistent between `disallow`, `ignore`, and `skip` everywhere

File: pkg/process/checks/process.go [hunk 6]
```diff
@@ -112,7 +116,7 @@ type ProcessCheck struct {
 func (p *ProcessCheck) Init(syscfg *SysProbeConfig, info *HostInfo, oneShot bool) error {
 	p.hostInfo = info
 	p.sysProbeConfig = syscfg
-	p.probe = newProcessProbe(p.config, procutil.WithPermission(syscfg.ProcessModuleEnabled))
+	p.probe = newProcessProbe(p.config, procutil.WithPermission(syscfg.ProcessModuleEnabled), procutil.WithIgnoreZombieProcesses(p.config.GetBool(configDisallowZombies)))
```
Comment (robertjli) [inline pkg/process/checks/process.go:119]: Optional: long lines can sometimes get hard to parse, we could break this up with a param on each line

File: pkg/process/procutil/process_linux.go [hunk 7]
```diff
@@ -250,6 +260,10 @@ func (p *probe) ProcessesByPID(now time.Time, collectStats bool) (map[int32]*Pro
 				//       no cmdline and they have the PF_KTHREAD flag set in /proc/<pid>/stat
 				//       Moving this check down the stack saves us from a number of needless follow-up system calls.
 				continue
+
+				// Check if toggle for ignoring zombies is on, and if it is ignore process
```
Comment (robertjli) [inline pkg/process/procutil/process_linux.go:264]: Nit, optional: I don't think this comment is necessary, we can just remove it

File: pkg/process/procutil/process_linux_test.go [hunk 8]
```diff
@@ -150,6 +150,11 @@ func TestProcessesByPIDTestFS(t *testing.T) {
 	testProcessesByPID(t, WithProcFSRoot("resources/test_procfs/proc/"))
 }
 
+func TestProcessesByPIDTestIgnoringZombiesFS(t *testing.T) {
+	t.Setenv("HOST_PROC", "resources/test_procfs/proc/")
+	testProcessesByPID(t, WithProcFSRoot("resources/test_procfs/proc/"), WithIgnoreZombieProcesses(true))
```
Comment (robertjli) [inline pkg/process/procutil/process_linux_test.go:155]: Nice, being able to reuse the test helper makes it super simple to test 👍

**Review by robertjli — COMMENTED** [review]
I think we should keep the logic in `fmtProcesses()`, because even though it does not protect the process agent from OOMing (which was the original reason for adding this option), it does still change the behavior of the agent in the way the customer expects when they turn on this config.

One last thing, I think this needs a changelog as the config change is customer-facing. You can read up on how to create one here: https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno

File: pkg/process/checks/process_test.go [hunk 9]
```diff
@@ -446,3 +446,127 @@ func BenchmarkProcessCheck(b *testing.B) {
...
+	processCheck.skipZombieProcesses = processCheck.config.GetBool(configDisallowZombies)
+
+	now := time.Now().Unix()
+	proc1 := makeProcessWithCreateTime(1, "git clone google.com", now)
+	proc2 := makeProcessWithCreateTime(2, "mine-bitcoins -all -x", now+1)
+	proc3 := makeProcessWithCreateTime(3, "foo --version", now+2)
+	proc4 := makeProcessWithCreateTime(4, "foo -bar -bim", now+3)
+	proc5 := makeProcessWithCreateTime(5, "datadog-process-agent --cfgpath datadog.conf", now+2)
+	proc4.Stats.Status = "Z"
+	proc5.Stats.Status = "Z"
```
Comment (daniel-taf) [inline pkg/process/checks/process_test.go:463]: Just a suggestion, but I think we can simplify this test and the other one a little bit by reducing the amount of processes you are creating. The amount of processes doesn't influence the behavior.

File: pkg/process/checks/process_test.go [hunk 10]
```diff
@@ -446,3 +446,99 @@ func BenchmarkProcessCheck(b *testing.B) {
...
+		&model.CollectorProc{
+			Processes: []*model.Process{expectedModel3},
+			GroupSize: int32(len(processesByPid)),
+			Info:      processCheck.hostInfo.SystemInfo,
+			Hints:     &model.CollectorProc_HintMask{HintMask: 0b1},
+		},
+	}
+	actual, err := processCheck.run(0, false)
+	require.NoError(t, err)
+	assert.Equal(t, len(expected), len(actual.Payloads()))
```
Comment (wiyu) [inline pkg/process/checks/process_test.go:499]: You can sort the values and assert against expected slice, otherwise you've created a unnecessary slice of expected processes.

Comment (robertjli) [reply pkg/process/checks/process_test.go]: You can also use `assert.ElementsMatch()`: https://pkg.go.dev/github.com/stretchr/testify/assert#ElementsMatch

Comment (levin0710) [reply pkg/process/checks/process_test.go]: @robertjli On the latest commit i changed it to `assert.ElementsMatch()`. Everything worked as expected

File: releasenotes/notes/ignore-zombie-process-config-cf78b9d48d34c702.yaml [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    Added config that allows user to toggle on and off the collection of Zombie processes in the agent.
```
Comment (cswatt) [inline releasenotes/notes/ignore-zombie-process-config-cf78b9d48d34c702.yaml:11]:
```suggestion
    Added config that allows user to toggle on and off the collection of zombie processes in the Agent.
```

File: releasenotes/notes/ignore-zombie-process-config-cf78b9d48d34c702.yaml [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+fixes:
+  - |
+    Prevention of OOMs due to excessive collection of Zombie processes.
```
Comment (cswatt) [inline releasenotes/notes/ignore-zombie-process-config-cf78b9d48d34c702.yaml:14]:
```suggestion
    Prevention of OOMs due to excessive collection of zombie processes.
```

**Review by cswatt — COMMENTED** [review]
Left spelling/capitalization feedback on release note.

**Review by robertjli — APPROVED** [review]

File: pkg/process/checks/process.go [hunk 13]
```diff
@@ -592,6 +602,9 @@ func skipProcess(
 		// This means short-lived processes (<2s) will never be captured.
 		return true
 	}
+	if zombiesIgnored && fp.Stats != nil && fp.Stats.Status == "Z" {
```
Comment (xornivore) [inline pkg/process/checks/process.go:605]: Nit: pretty obvious by reading the code, but may be helpful to add a small comment and reference procfs doc here.

File: releasenotes/notes/ignore-zombie-process-config-cf78b9d48d34c702.yaml (hunk 11 above)
Comment (xornivore) [reply releasenotes/notes/ignore-zombie-process-config-cf78b9d48d34c702.yaml:11]: We may want to change this to Process Agent.

File: releasenotes/notes/ignore-zombie-process-config-cf78b9d48d34c702.yaml (hunk 12 above)
Comment (xornivore) [reply releasenotes/notes/ignore-zombie-process-config-cf78b9d48d34c702.yaml:14]: Nit: "excessive collection" sounds off to me. Maybe "Prevention of OOMs when collecting a large number of zombie processes."

**Review by xornivore — COMMENTED** [review]
LGTM overall but let's document in the PR description that the new setting is not set to `true`. Sanity check if `true` is ever going to be the default? Do we expect users to know about it and configure or the default of `true` is what they want?

File: pkg/process/checks/process.go [hunk 14]
```diff
@@ -69,6 +70,9 @@ type ProcessCheck struct {
 	// disallowList to hide processes
 	disallowList []*regexp.Regexp
 
+	// determine if zombies process will be collected
+	ignoreZombieProcesses bool
```
Comment (xornivore) [inline pkg/process/checks/process.go:74]: Per @wiyu prior comment - should this be `skipZombieProcesses` to be more inline with the naming of the `skipProcess`, then it just makes it trivial to understand.

File: pkg/process/checks/process.go [hunk 15]
```diff
@@ -430,11 +438,12 @@ func fmtProcesses(
 	connRates ProcessConnRates,
 	//nolint:revive // TODO(PROC) Fix revive linter
 	lookupIdProbe *LookupIdProbe,
+	zombiesIgnored bool,
```
Comment (xornivore) [inline pkg/process/checks/process.go:441]: Nit: any reason not to rename this to `skipZombieProcesses`?

**Review by daniel-taf — APPROVED** [review]

**Comment by levin0710** [comment]
> LGTM overall but let's document in the PR description that the new setting is not set to `true`. Sanity check if `true` is ever going to be the default? Do we expect users to know about it and configure or the default of `true` is what they want?

PR description was updated and provides clarification on this information.

File: pkg/process/checks/process.go (hunk 15 above)
Comment (levin0710) [reply pkg/process/checks/process.go:441]: For consistency purposes, decided to keep it as `ignoreZombieProcesses`

**Review by xornivore — APPROVED** [review]

**Review by cswatt — APPROVED** [review]
